File: src/multigrid.cpp
#include "multigrid.h"

/**
 * Using Gauss-Seidel method as pre- and post-smoothing to reduce high frequency errors.
 */
void smooth(Grid& psi, const Grid& rhs, int iterations, double h) {
    int n = psi.size();
    Grid psi_new = zero_grid(n);
    for (int it = 0; it < iterations; ++it) {
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                psi_new[i][j] = 0.25 * (psi[idx(i+1, n)][j] + psi[idx(i-1, n)][j]
                            + psi[i][idx(j+1, n)] + psi[i][idx(j-1, n)]
                            - h * h * rhs[i][j]);
            }
        }
        psi = psi_new;
    }
}

/**
 * Compute residual errors ：res = rhs - Lψ
 */
Grid residual(const Grid& psi, const Grid& rhs, double h) {
    return rhs - laplacian(psi, h);
}

/**
 * Downsampling (restricting) the residual error to a coarser grid.
 */
Grid restrict_grid(const Grid& fine) {
    int Nc = fine.size() / 2;
    int n = fine.size();
    Grid coarse = zero_grid(Nc);
    for (int i = 0; i < Nc; ++i) {
        for (int j = 0; j < Nc; ++j) {
            coarse[i][j] = 0.25 * (fine[idx(2*i, n)][idx(2*j, n)]
                                    + fine[idx(2*i+1, n)][idx(2*j, n)]
                                    + fine[idx(2*i, n)][idx(2*j+1, n)]
                                    + fine[idx(2*i+1, n)][idx(2*j+1, n)]);
        }
    }
    return coarse;
}
/**
 * Interpolating (prolonging) a correction computed on a coarser grid into a finer grid.
 */
Grid prolong_grid(const Grid& coarse) {
    int Nc = coarse.size();
    int Nf = Nc * 2;
    Grid fine = zero_grid(Nf);
    for (int i = 0; i < Nc; ++i) {
        for (int j = 0; j < Nc; ++j) {
            for (int di = 0; di < 2; ++di) {
                for (int dj = 0; dj < 2; ++dj) {
                    fine[idx(2*i + di, Nf)][idx(2*j + dj, Nf)] = coarse[i][j];
                }
            }
        }
    }
    return fine;
}

/**
 *  V-cycle Multigrid main function
    - level = 0 means the coarsest gird
    - Otherwise, recursively downsampling, computing errors and prolonging
 */
Grid multigrid_vcycle(const Grid& rhs, int level, double h) {
    int n = rhs.size();
    Grid psi = zero_grid(n);

    // Stop recursion at smallest grid size, otherwise continue recursion
    if (n <= 4 || level == 0) {
        smooth(psi, rhs, 50, h);
        return psi;
    }

    // Pre-Smoothing
    smooth(psi, rhs, 3, h);
    // Compute Residual Errors
    Grid r = residual(psi, rhs, h);
    // Restriction
    Grid r_coarse = restrict_grid(r);
    // Recursively compute errors
    Grid e_coarse = multigrid_vcycle(r_coarse, level - 1, 2 * h);
    // Prolongation
    Grid e_fine = prolong_grid(e_coarse);
    // Correction
    psi = psi + e_fine;
    // Post-Smoothing
    smooth(psi, rhs, 3, h);

    return psi;
}

File: post_processing/multigrid.py
import numpy as np

def laplacian(f, dx):
    return (
        np.roll(f, 1, axis=0) + np.roll(f, -1, axis=0) +
        np.roll(f, 1, axis=1) + np.roll(f, -1, axis=1) - 4*f
    ) / dx**2

def smooth(psi, rhs, dx, iterations=3):
    """
    Using Gauss-Seidel method as pre- and post-smoothing to reduce high frequency
    errors.
    """
    for _ in range(iterations):
        psi = 0.25 * (
            np.roll(psi, 1, axis=0) + np.roll(psi, -1, axis=0) +
            np.roll(psi, 1, axis=1) + np.roll(psi, -1, axis=1) -
            dx**2 * rhs
        )
    return psi

def residual(psi, rhs, dx):
    # Compute residual errors ：res = rhs - Lψ
    return rhs - laplacian(psi, dx)

def restrict(fine):
    """
    Downsampling (restricting) the residual error to a coarser grid.
    """
    Nc = fine.shape[0] // 2
    coarse = np.zeros((Nc, Nc))
    for i in range(Nc):
        for j in range(Nc):
            ii, jj = 2 * i, 2 * j
            coarse[i, j] = 0.25 * (fine[ii, jj] + fine[ii+1, jj] +
                                   fine[ii, jj+1] + fine[ii+1, jj+1])
    return coarse

def prolong(coarse):
    """
    Interpolating (prolonging) a correction computed on a coarser grid into a finer grid.
    """
    Nc = coarse.shape[0]
    Nf = Nc * 2
    fine = np.zeros((Nf, Nf))
    for i in range(Nc):
        for j in range(Nc):
            fine[2*i:2*i+2, 2*j:2*j+2] = coarse[i, j]
    return fine

def v_cycle(rhs, dx, level):
    """
    V-cycle Multigrid main function
    - level = 0 means the coarsest gird
    - Otherwise, recursively downsampling, computing errors and prolonging
    """
    N = rhs.shape[0]
    psi = np.zeros_like(rhs)

    # Stop recursion at smallest grid size, otherwise continue recursion
    if level == 0 or N <= 4:
        return smooth(psi, rhs, dx, iterations=50)

    # Pre-Smoothing
    psi = smooth(psi, rhs, dx)
    # Compute Residual Errors
    res = residual(psi, rhs, dx)
    # Restriction
    res_c = restrict(res)
    # Recursively compute errors
    err_c = v_cycle(res_c, 2*dx, level-1)
    # Prolongation and Correction
    err_f = prolong(err_c)
    psi += err_f
    # Post-Smoothing
    psi = smooth(psi, rhs, dx)

    return psi

def w_cycle(rhs, dx, level):
    """
    W-cycle Multigrid main function
    - level = 0 means the coarsest grid
    - Otherwise, recursively apply two coarse-level corrections
    """
    N = rhs.shape[0]
    psi = np.zeros_like(rhs)

    # Stop recursion at coarsest grid
    if level == 0 or N <= 4:
        return smooth(psi, rhs, dx, iterations=50)

    # Pre-smoothing
    psi = smooth(psi, rhs, dx)

    # Compute residual
    res = residual(psi, rhs, dx)

    # Restrict to coarse grid
    res_c = restrict(res)

    # First coarse-grid correction
    err_c1 = w_cycle(res_c, 2*dx, level - 1)

    # Second coarse-grid correction (W-cycle key step)
    err_c2 = w_cycle(err_c1, 2*dx, level - 1)

    # Prolongate and apply correction
    err_f = prolong(err_c2)
    psi += err_f

    # Post-smoothing
    psi = smooth(psi, rhs, dx)

    return psi

def solve_poisson(zeta, dx):
    # Given ζ，solve Δψ = ζ
    rhs = zeta
    levels = int(np.log2(zeta.shape[0])) - 2
    return v_cycle(rhs, dx, levels)
    #return w_cycle(rhs, dx, levels)